sdks/go/pkg/beam/io/spannerio/read_batch.go

```diff
@@ -0,0 +1,168 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spannerio
+
+import (
+	"context"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/sdf"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/io/rtrackers/offsetrange"
+	"reflect"
+
+	"cloud.google.com/go/spanner"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/register"
+	"google.golang.org/api/iterator"
+)
+
+func init() {
+	register.DoFn4x1[context.Context, *sdf.LockRTracker, PartitionedRead, func(beam.X), error]((*readBatchFn)(nil))
+	register.Emitter1[beam.X]()
+}
+
+type readBatchFn struct {
+	spannerFn
+	Type    beam.EncodedType
+	Options queryOptions
+}
+
+func newReadBatchFn(db string, t reflect.Type, options queryOptions) *readBatchFn {
+	return &readBatchFn{
+		spannerFn: newSpannerFn(db),
+		Type:      beam.EncodedType{T: t},
+		Options:   options,
+	}
+}
+
+func readBatch(s beam.Scope, db string, query string, t reflect.Type, options queryOptions) beam.PCollection {
+	partitions := generatePartitions(s, db, query, t, options)
+
+	s = s.Scope("spannerio.ReadBatch")
+
+	return beam.ParDo(
+		s,
+		newReadBatchFn(db, t, options),
+		partitions,
+		beam.TypeDefinition{Var: beam.XType, T: t},
+	)
+}
+
+func (f *readBatchFn) Setup(ctx context.Context) error {
+	return f.spannerFn.Setup(ctx)
+}
+
+// CreateInitialRestriction creates an offset range restriction representing
+// the partition's size in bytes.
+func (f *readBatchFn) CreateInitialRestriction(read PartitionedRead) offsetrange.Restriction {
+	txn := f.client.BatchReadOnlyTransactionFromID(read.BatchTransactionId)
+	iter := txn.Execute(context.Background(), read.Partition)
+	defer iter.Stop()
+
+	return offsetrange.Restriction{
+		Start: 0,
+		End:   iter.RowCount,
+	}
+}
+
+const (
+	blockSize = 10000
+	tooSmall  = 100
+)
+
+// SplitRestriction splits each file restriction into blocks of a predetermined
+// size, with some checks to avoid having small remainders.
+func (f *readBatchFn) SplitRestriction(_ PartitionedRead, rest offsetrange.Restriction) []offsetrange.Restriction {
+	splits := rest.SizedSplits(blockSize)
```

Comment:
Note that this approach also may have the same issue as textIO does with a fixed size split, see #25892. However, since Spanner has already done a bunch of splits for us, it might be moot.  Not required to fix this now, but just a note.